### PR TITLE
[feature] Local environment setup

### PR DESCRIPTION
**Pull request by xBalbinus**
## Motivation

This is a followup to the local environment RFC https://doc.clickup.com/14205595/d/h/dhgmv-10480/8b498d6fedd36f7, which should help contributors and developers set up local testing environments for Nomad.

## Solution

[The RFC Spec listed here](https://doc.clickup.com/14205595/d/h/dhgmv-10480/8b498d6fedd36f7)

Changelog:
- Restored and cleaned up previously broken contract and agent deployment abstractions (it works now!)
- Dockerized hardhat and agent image are now ready to use for further integration tests (integration tests themselves are WIP and coming in a later ticket)
- Yarn bootup / yarn start scripts ready for booting up the local environment (further yarn scripts for individual integration tests to be implemented in following PR)
- Removed jokes on variables, set up changelog to track oncoming additions and changes
- Added Bunyan for logging capabilities

Testing:
- Tested multiple times on local machine, with the help of the tooling team @yourbuddyconner @kekonen @luketchang @Imti et. al.
- Documentation and .env.examples updated to fit the currently running setup.

## PR Checklist

- [x] Added Tests
- [x] Updated Documentation
- [x] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

**Comment by xBalbinus** [comment]
p.s. Please don't mind the merge conflicts those are not intentional and will be resolved.

**Comment by anna-carroll** [comment]
@xBalbinus @yourbuddyconner is this PR ready for review, or should be marked as draft?

if ready, can you fix build & let's get it reviewed!

File: packages/contracts-core/contracts/governance/GovernanceRouter.sol [hunk 1]
```diff
@@ -346,6 +346,7 @@ contract GovernanceRouter is Version0, Initializable, IMessageRecipient {
      * dispatch the change to all remote routers
      * @param _domain The domain
      * @param _router The address of the new router
+     * Router domain NOT GETTING SET PROPERLY FOR JERRY.
```
Comment (yourbuddyconner) [inline packages/contracts-core/contracts/governance/GovernanceRouter.sol:349]: Remove comments from non-local-environment packages

File: packages/contracts-core/contracts/governance/GovernanceRouter.sol [hunk 2]
```diff
@@ -545,6 +546,7 @@ contract GovernanceRouter is Version0, Initializable, IMessageRecipient {
         // require that the new governor is not the zero address
         require(_newGovernor != address(0), "cannot renounce governor");
         // require that the governor domain has a valid router
+        // returns false on jerry because governor is 0x0.
```
Comment (yourbuddyconner) [inline packages/contracts-core/contracts/governance/GovernanceRouter.sol:549]: Same here

File: packages/deploy/src/DeployContext.ts [hunk 3]
```diff
@@ -138,10 +138,13 @@ export class DeployContext extends MultiProvider<config.Domain> {
     verification: Verification,
   ): void {
     const name = this.resolveDomainName(nameOrDomain);
+    // console.log(name);
```
Comment (yourbuddyconner) [inline packages/deploy/src/DeployContext.ts:141]: Clean up

File: packages/deploy/src/bridge/BridgeContracts.ts [hunk 4]
```diff
@@ -122,7 +122,7 @@ export default class BridgeContracts extends AbstractBridgeDeploy<config.EvmBrid
     const factory = new contracts.BridgeToken__factory(this.deployer);
     const impl = await factory.deploy(this.overrides);
     await impl.deployTransaction.wait(this.confirmations);
-
+
```
Comment (yourbuddyconner) [inline packages/deploy/src/bridge/BridgeContracts.ts:125]: Revert non-substantive changes like this

File: packages/deploy/src/core/CoreContracts.ts [hunk 5]
```diff
@@ -703,6 +703,8 @@ export default class EvmCoreDeploy extends AbstractCoreDeploy<config.EvmCoreCont
       );
       await tx.wait(this.confirmations);
     }
+
+    console.log(this.context.verification);
```
Comment (yourbuddyconner) [inline packages/deploy/src/core/CoreContracts.ts:707]: Revert this

File: packages/local-environment/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+not have to worry about including a link manually, but you can if you wish.
+
+Types of changes (Stanzas):
+
+"Features" for new features.
+"Improvements" for changes in existing functionality.
+"Deprecated" for soon-to-be removed features.
+"Bug Fixes" for any bug fixes.
+"Client Breaking" for breaking CLI commands and REST routes used by end-users.
+Ref: https://keepachangelog.com/en/1.0.0/
```
Comment (yourbuddyconner) [inline packages/local-environment/CHANGELOG.md:29]: Love this

File: packages/local-environment/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+## Bootup instructions
+
```
Comment (yourbuddyconner) [inline packages/local-environment/README.md:2]: Mention these commands should be run from monorepo root

File: packages/local-environment/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+PRIVATE_KEY_2=2000000000000000000000000000000000000000000000000000000000000002
+PRIVATE_KEY_3=3000000000000000000000000000000000000000000000000000000000000003
+PRIVATE_KEY_4=4000000000000000000000000000000000000000000000000000000000000004
+PRIVATE_KEY_5=5000000000000000000000000000000000000000000000000000000000000005
+```
+
+Then, boot up docker.
+
+Run the following inside /packages/local-environment if this is your first time booting up.
+
```
Comment (yourbuddyconner) [inline packages/local-environment/README.md:25]: Mention this should be run from local-environment package dir

File: packages/local-environment/hardhat/hardhat.config.js [hunk 9]
```diff
@@ -1,4 +1,5 @@
 require("@nomiclabs/hardhat-waffle");
+require("@nomiclabs/hardhat-etherscan");
```
Comment (yourbuddyconner) [inline packages/local-environment/hardhat/hardhat.config.js:2]: Why is this needed again?

File: packages/local-environment/hardhat/hardhat.config.js [hunk 10]
```diff
@@ -10,19 +11,21 @@ task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
...
+    process.env.PRIVATE_KEY ||
     "1337000000000000000000000000000000000000000000000000000000001337",
 
+
+    process.env.PRIVATE_KEY1 || '1000000000000000000000000000000000000000000000000000000000000001',
+    process.env.PRIVATE_KEY2 || '2000000000000000000000000000000000000000000000000000000000000002',
+    process.env.PRIVATE_KEY3 || '3000000000000000000000000000000000000000000000000000000000000003',
+    process.env.PRIVATE_KEY4 || '4000000000000000000000000000000000000000000000000000000000000004',
+    process.env.PRIVATE_KEY5 || '5000000000000000000000000000000000000000000000000000000000000005',
+
```
Comment (yourbuddyconner) [inline packages/local-environment/hardhat/hardhat.config.js:24]: Lets include support for 10 private keys here

File: packages/local-environment/hardhat/hardhat.config.js [hunk 11]
```diff
@@ -10,19 +11,21 @@ task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
...
   ...Object.entries(process.env)
     .filter(([k, _]) => k.match(/PRIVATE_KEY\d+/))
     .map(([_, v]) => v),
 
-  // process.env.PRIVATE_KEY1 || '1000000000000000000000000000000000000000000000000000000000000001',
-  // process.env.PRIVATE_KEY2 || '2000000000000000000000000000000000000000000000000000000000000002',
-  // process.env.PRIVATE_KEY3 || '3000000000000000000000000000000000000000000000000000000000000003',
-  // process.env.PRIVATE_KEY4 || '4000000000000000000000000000000000000000000000000000000000000004',
-  // process.env.PRIVATE_KEY5 || '5000000000000000000000000000000000000000000000000000000000000005',
   // process.env.PRIVATE_KEY6 || '6000000000000000000000000000000000000000000000000000000000000006',
```
Comment (yourbuddyconner) [inline packages/local-environment/hardhat/hardhat.config.js:29]: Remove the rest of the comments here

File: packages/local-environment/src/agent.ts [hunk 12]
```diff
@@ -20,11 +35,11 @@ export interface Agent {
 }
 
 export enum AgentType {
-  Updater,
-  Relayer,
-  Processor,
-  Watcher,
-  Kathy,
+  Updater = 'updater',
```
Comment (yourbuddyconner) [inline packages/local-environment/src/agent.ts:38]: Is there a reason you set values to the enum members here? I would roll this back unless there's an explicit thing this is fixing.

Comment (yourbuddyconner) [reply packages/local-environment/src/agent.ts:38]: EDIT: Well I see this enabled you to remove `agentTypeToString` which is an improvement I think

Comment (yourbuddyconner) [reply packages/local-environment/src/agent.ts:38]: Would maybe ask @prestwich if he has any hot-takes around the philosophically best way to handle this, I think adding literal values to the enum is a good solution right now.

File: packages/local-environment/src/CLI/bootup.sh [hunk 13]
```diff
@@ -0,0 +1,5 @@
+pushd hardhat
```
Comment (yourbuddyconner) [inline packages/local-environment/src/CLI/bootup.sh:1]: Put this in package root

File: packages/local-environment/src/network.ts [hunk 14]
```diff
@@ -1,359 +1,596 @@
...
-      },
-      HostConfig: {
-        PortBindings: {
-          "8545/tcp": [
-            {
-              HostPort: port.toString(),
+export abstract class Network {
+    agents?: Agents;
+
+    connectedNetworks: Network[];
```
Comment (yourbuddyconner) [inline packages/local-environment/src/network.ts:22]: Thinking we should move the concept of "connections" to the higher-level abstraction

File: packages/local-environment/src/network.ts [hunk 15]
```diff
@@ -1,359 +1,596 @@
...
+
+    name: string;
+    domainNumber: number;
+    chainId: number;
+    deployOverrides: ethers.Overrides;
+    signers: Map<number | string, Key>;
+    updaters: Map<number | string, Key>;
```
Comment (yourbuddyconner) [inline packages/local-environment/src/network.ts:29]: Think we should move any concept of "agents" to the higher-level abstraction, Network should be pure and explicitly encapsulate blockchain primitives like keys, accounts, and blockchain constants

File: packages/local-environment/src/network.ts [hunk 16]
```diff
@@ -1,359 +1,596 @@
...
+    watchers: Map<number | string, Key>;
+    relayers: Map<number | string, Key>;
+    kathys: Map<number | string, Key>;
+    processors: Map<number | string, Key>;
+
+    coreContracts?: CoreContracts;
```
Comment (yourbuddyconner) [inline packages/local-environment/src/network.ts:35]: Same here, no Nomad things in Network

File: packages/local-environment/src/network.ts [hunk 17]
```diff
@@ -1,359 +1,596 @@
...
+        const wethFactory = new ethers.ContractFactory(wethABI, wethByteCode, owner);
+        const weth =  await wethFactory.deploy();
+        const wethAddyTom = await weth.address;
```
Comment (yourbuddyconner) [inline packages/local-environment/src/network.ts:523]: Remove reference to tom

**Review by yourbuddyconner — CHANGES_REQUESTED** [review]

File: packages/contracts-core/contracts/governance/GovernanceRouter.sol (hunk 1 above)
Comment (xBalbinus) [reply packages/contracts-core/contracts/governance/GovernanceRouter.sol]: Fixed & reviewing!

File: packages/contracts-core/contracts/governance/GovernanceRouter.sol (hunk 2 above)
Comment (xBalbinus) [reply packages/contracts-core/contracts/governance/GovernanceRouter.sol]: Fix'd

File: packages/deploy/src/DeployContext.ts (hunk 3 above)
Comment (xBalbinus) [reply packages/deploy/src/DeployContext.ts]: Cleaned

File: packages/deploy/src/bridge/BridgeContracts.ts (hunk 4 above)
Comment (xBalbinus) [reply packages/deploy/src/bridge/BridgeContracts.ts]: Reverted

File: packages/deploy/src/core/CoreContracts.ts (hunk 5 above)
Comment (xBalbinus) [reply packages/deploy/src/core/CoreContracts.ts]: Reverted

File: packages/local-environment/README.md (hunk 7 above)
Comment (xBalbinus) [reply packages/local-environment/README.md:2]: Done

File: packages/local-environment/README.md (hunk 8 above)
Comment (xBalbinus) [reply packages/local-environment/README.md:29]: Done

File: packages/local-environment/hardhat/hardhat.config.js (hunk 11 above)
Comment (xBalbinus) [reply packages/local-environment/hardhat/hardhat.config.js]: removed & keys support added

File: packages/local-environment/hardhat/hardhat.config.js (hunk 9 above)
Comment (xBalbinus) [reply packages/local-environment/hardhat/hardhat.config.js:2]: Hmm I don't actually think that's necessary -  was just there. Removing.

File: packages/local-environment/src/CLI/bootup.sh (hunk 13 above)
Comment (xBalbinus) [reply packages/local-environment/src/CLI/bootup.sh]: Relocated.

File: packages/local-environment/src/network.ts (hunk 17 above)
Comment (xBalbinus) [reply packages/local-environment/src/network.ts]: Removed

File: packages/local-environment/CHANGELOG.md (hunk 6 above)
Comment (kekonen) [reply packages/local-environment/CHANGELOG.md:29]: +++

File: packages/local-environment/src/agent.ts (hunk 12 above)
Comment (kekonen) [reply packages/local-environment/src/agent.ts:38]: Doing it the Xiangan's way is definitely correct. Writing this I was not aware of TS' way of assigning strings to enums

File: packages/local-environment/src/network.ts (hunk 15 above)
Comment (kekonen) [reply packages/local-environment/src/network.ts]: I agree with Conner on that one. The point is to make a network an independent actor. Idea: 
have a structure NomadEnv have multiple Domains, where Domain will have Keys, one Network(actor) and multiple Agents(actors).

**Comment by Imti** [comment]
Closing since no longer needed
